ddtrace/contrib/django/apps.py

```diff
@@ -36,8 +37,8 @@ def ready(self):
         # AgentWriter, it breaks all tests. The configure() behavior must
         # be changed to use it in this integration
         tracer.enabled = settings.ENABLED
-        tracer.writer.api.hostname = settings.AGENT_HOSTNAME
-        tracer.writer.api.port = settings.AGENT_PORT
+        tracer.writer.api.hostname = os.environ.get("DATADOG_TRACE_AGENT_HOSTNAME") or settings.AGENT_HOSTNAME
```

Comment:
It's definitely an import change! But what about making the change here: https://github.com/spesnova/dd-trace-py/blob/8b677483352c81cdaf773d1ba273f76d538d65cd/ddtrace/contrib/django/conf.py#L25-L26 ?

What I mean is that I prefer ENV < hard-coded settings, so that `DATADOG_TRACE ` settings are not overridden by env variables if set. In that way, you can just:
```python
DEFAULTS = {
    'AGENT_HOSTNAME': os.environ.get("DATADOG_TRACE_AGENT_HOSTNAME", 'localhost'),
    'AGENT_PORT': os.environ.get("DATADOG_TRACE_AGENT_PORT", 8126),
    # ...
}
```

This translates in having defaults as env vars, unless specified otherwise. 

Also, consider that `tracer.writer.api.port` must be an `int` otherwise the connection string will not work. Reference this PR: https://github.com/DataDog/dd-trace-py/pull/140

Does it make sense? What do you think?

Reply:
> Also, consider that tracer.writer.api.port must be an int otherwise the connection string will not work. Reference this PR: #140

OK. I'll change here!

Reply:
> What I mean is that I prefer ENV < hard-coded settings, so that DATADOG_TRACE settings are not overridden by env variables if set. In that way, you can just:
> This translates in having defaults as env vars, unless specified otherwise.
> Does it make sense? What do you think?
 
I have opposite idea. This mean I want to use `DATADOG_TRACE` settings as default value, and override it with ENV if needed.

Let me explain the reason. This pull request is originally come from a problem when running Django (and dd-trace-py) in a Docker container. Most Django apps are stateless application. So I want to apply Heroku's "12-Factor App" to my application. They say "Store config in the environment":

> Apps sometimes store config as constants in the code. This is a violation of twelve-factor, which requires strict separation of config from code. Config varies substantially across deploys, code does not.

I understand that we don't have to store every config in environment. 

> Note that this definition of “config” does not include internal application config, such as config/routes.rb in Rails, or how code modules are connected in Spring. This type of config does not vary between deploys, and so is best done in the code.

I think agent host and agent port are type of config which should be separated from code.

Please read ["Config" part](https://12factor.net/config) for more details about 12-Factor App.

Reply:
What I suggested is relying in your environment variables by default, so until developers hard code settings, that works like you're saying and it should respect 12-Factor App configuration (so `os.environ` is used without writing anything in your Django settings).

In that way if you put the Django app in a container without writing APM settings, and you pass env variables, it will work as you're expecting. On the other hand, if you hard code the values because you want to programmatically force something, env variables will be ignored. That case is not 12-Factor compliant but users may expect that if you hard-code something, it will be used with highest priority. So with my suggestion, just don't hard-code anything and it would pick your env variables.

Am I missing something @spesnova?

Reply:
I understand what you suggested now.

In Rails world, if I have both hard-coded and environment variable set, the environment variable will take precedence. Since I'm used to Rails, I thought environment variable should take precedence this time too.

http://guides.rubyonrails.org/configuring.html#connection-preference

I think there is no "one right answer" here and this pull request's main purpose is changing django integration to respect env-vars. So I'll update code due to what you suggested 👍

Reply:
Yes I think so; a right answer is difficult on such topic since it depends on framework / language conventions. Anyway, what I really want is being sure that with my suggestion your Django application works as you expect. Happy to review and then merge your PR!

Thank you for taking time investigating and providing very good references!